Userland/Libraries/LibGfx/PNGWriter.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2021, Pierre Hoffmeister
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/String.h>
+#include <AK/Vector.h>
+#include <LibCore/FileStream.h>
+#include <LibGfx/Bitmap.h>
+
+namespace Gfx {
+
+class PNGChunk {
+public:
+    PNGChunk(const String&);
+    const Vector<u8>& data() { return m_data; };
+    const String& type() { return m_type; };
+    void add_u8(u8);
+    void add_u16_big(u16);
+    void add_u32_big(u32);
+    void add_u16_little(u16);
+    void add_u32_little(u32);
+
+private:
+    Vector<u8> m_data;
+    String m_type;
+};
+
+class NonCompressibleBlock {
+public:
+    NonCompressibleBlock()
+        : m_adler_s1(1)
+        , m_adler_s2(0)
+    {
+    }
+
+    void finalize(PNGChunk&);
+    void add_byte_to_block(u8 data, PNGChunk&);
+
+    u32 adler_s1() { return m_adler_s1; }
+    u32 adler_s2() { return m_adler_s2; }
+
+private:
+    void add_block_to_chunk(PNGChunk&, bool);
+    void clear() { m_non_compressible_data.clear(); };
+    void update_adler(u8);
+    bool full() { return m_non_compressible_data.size() == 65535; }
+    Vector<u8> m_non_compressible_data;
+    u32 m_adler_s1;
+    u32 m_adler_s2;
```

Comment:
I don't think this class needs a non-trivial constructor.
```suggestion
    void finalize(PNGChunk&);
    void add_byte_to_block(u8 data, PNGChunk&);

    u32 adler_s1() { return m_adler_s1; }
    u32 adler_s2() { return m_adler_s2; }

private:
    void add_block_to_chunk(PNGChunk&, bool);
    void clear() { m_non_compressible_data.clear(); };
    void update_adler(u8);
    bool full() { return m_non_compressible_data.size() == 65535; }
    Vector<u8> m_non_compressible_data;
    u32 m_adler_s1 { 1 };
    u32 m_adler_s2 { 0 };
```

Reply:
Done. :+1: